### ISSUE TITLE
Suboptimal performance on large buffers (sftp, stdin)

### ISSUE THREAD

**Issue by jkji** [paramiko/paramiko#892]
The following lines lead to suboptimal performance on large buffers.
A single `_write` or `send` typically sends ~32K bytes, while copying most of the remaining `data` passed in e.g. `sftp.file.write(data)` to a new buffer. If `data` is a few hundred MB, this can cap the transfer rate with 100% cpu usage.

Sftp: https://github.com/paramiko/paramiko/blob/master/paramiko/file.py#L499

```python
while len(data) > 0:
    count = self._write(data)
    data = data[count:]
```

Stdin: https://github.com/paramiko/paramiko/blob/master/paramiko/channel.py#L814

```python
while s:
    sent = self.send(s)
    s = s[sent:]
```

Sample benchmark on a 100MB buffer:

```
$ time python -c "a='0'*(100*1024*1024)
while a: a=a[32700:]"

21.92s user 15.02s system 99% cpu 36.982 total
```

**Comment by bz2** [comment]
It's possible that with #971 passing a `memoryview` in will improve performance for some paths.

**Comment by bitprophet** [comment]
Should be fixed in b4bd81dce17d23b3e402b7fd492bb2ebd30b284c
